### PR TITLE
Do multiple shooting

### PR DESCRIPTION
**Pull request by Manas2030**
This is in reference to issue #481

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Needs to be `include`d at the top level, needs docs.

File: src/multiple_shooting.jl [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+- `tsteps`: Timesteps on which ode_data was calculated.
+- `prob`: ODE problem that the Neural Network attempts to solve.
+- `loss_function`: Any arbitrary function to calculate loss.
+- `grp_size`: The group size achieved after splitting the ode_data into equal sizes.
+- `continuity_term`: Multiplying factor to ensure continuity of predictions throughout different groups.
+
+!!!note
+		The parameter 'continuity_term' should be a relatively big number to enforce a large penalty whenever the last point of any group doesn't coincide with the first point of next group.
+"""
+function loss_neuralode_helper(p::Array,ode_data::Array,tsteps::Array,prob::ODEProblem,loss_function::Function,grp_size::Integer=5,continuity_term::Integer=100)
```
Comment (Vaibhavdixit02) [inline src/multiple_shooting.jl:22]: This should probably have "multiple shooting" in its name. Also, have you taken a look at https://github.com/SciML/DiffEqParamEstim.jl/blob/e99fb8233f48be99540dc9bf63ac226bb22c3c22/src/multiple_shooting_objective.jl#L24?

Comment (Manas2030) [reply src/multiple_shooting.jl:22]: Hi, I'll change the name of the function to 'multiple_shooting'. And I took a look at the link now, do you have any suggestions from there? 😅

**Comment by yewalenikhil65** [comment]
I am little new to these concepts. Can anyone please tell me how is `mini-batching` different from `multiple-shooting` ? I could gather that both involve chunks of datasets to be trained, but i couldn't see much difference.

**Comment by ChrisRackauckas** [comment]
you don't solve the whole interval, and you do multiple batches in a loss call

**Comment by yewalenikhil65** [comment]
> you don't solve the whole interval, and you do multiple batches in a loss call

ohh yes..So,
1) loss is summed up over all the batches in multiple-shooting, 
2) initial condition is also changed accordingly in multiple-shooting..unlike in mini-batching

**Comment by ChrisRackauckas** [comment]
yes

File: src/multiple_shooting.jl (hunk 1 above)
Comment (Vaibhavdixit02) [reply src/multiple_shooting.jl:22]: Not really, just added it for your reference

File: src/multiple_shooting.jl [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+		pred = solve(remake(prob, p = p, tspan = (tsteps[1],tsteps[datasize]), u0 = ode_data[:,1]), Tsit5(),saveat = tsteps)
+		return tot_loss, pred
+	end
+
+	for i in 1:grp_size-1:datasize-grp_size
+		pred = solve(remake(prob, p = p, tspan = (tsteps[i],tsteps[i+grp_size-1]), u0 = ode_data[:,i]), Tsit5(),saveat = tsteps[i:i+grp_size-1])
+		tot_loss += loss_function(ode_data[:,i:i+grp_size-1], pred[:,1:grp_size]) + (continuity_term * sum(abs,pred[:,grp_size] - ode_data[:,i+grp_size-1]))
+	end
+	pred = solve(remake(prob, p = p, tspan = (tsteps[1],tsteps[datasize]), u0 = ode_data[:,1]), Tsit5(),saveat = tsteps)
```
Comment (yewalenikhil65) [inline src/multiple_shooting.jl:39]: Shouldn't the `Line 39`, be a concatenation of all the `pred` calculated inside the `for` loops, which is actually prediction corresponding to the `tot_loss` ?

**Review by yewalenikhil65 — COMMENTED** [review]
Shouldn't the Line 39, be a concatenation of all the pred calculated inside the for loops, which is actually prediction corresponding to the tot_loss ?

**Comment by Manas2030** [comment]
> Shouldn't the `Line 39`, be a concatenation of all the `pred` calculated inside the `for loops`, which is actually prediction corresponding to the `tot_loss` ?

Yes it indeed would be the actual predictions but I assume multiple shooting is just for training (i.e. calculating the loss) and the actual predictions (that is plotting the curve) would be done in a single shoot. Also the end points in any group overlap so there are 2 different predictions for them which would be weird to plot(?).

@ChrisRackauckas  What are your thoughts on this?

**Comment by yewalenikhil65** [comment]
> > Shouldn't the `Line 39`, be a concatenation of all the `pred` calculated inside the `for loops`, which is actually prediction corresponding to the `tot_loss` ?
> 
Also the end points in any group overlap so there are 2 different predictions for them which would be weird to plot(?).
> 
1-2, 2-3..say these are the batches
1) wrong prediction at `2` while you train from `1-2`
2) but then this `prediction` at `2` will be rectified because you use the `ode_data` at `2` initial condition for `2-3`

I think, the in between points prediction would be  accurate, because that `endpoint` is your initial condition for the next prediction... only intermediate points within the intervals would be off the exact `ode_data`.. Correct ?

**Comment by yewalenikhil65** [comment]
> Yes, I considered this and it's true we'll only have to overwrite end predictions of groups but while calculating loss we do sum up the offset of the point that we are overwriting. So it makes sense to display that offset as well if we are going to display it for the other remaining points.
> 
> Also, this will bring up discontinuties while plotting (which is fine but depends more on what the user would want)
> 
> And for the example you gave, it would plot the exact data we have, i.e. `ode_data`, if we overwrite each of the overlapping points (since every point is overlapping). So graphically our loss would always be 0 but this shouldn't happen

correct.. so i suggest this
**What seems then more correct to me is, we should concatenate all pred calculated inside for loop, but should drop the pred[,:1] for all i=2,3,4....n as they are just initial conditions.**

This won't bring discontinuities, and also it won't create `loss` to be zero prematurely.

**Comment by Vaibhavdixit02** [comment]
I don't recall if there was a discussion around this before, you should go through the related commits/PRs @Manas2030 but it does look like were returning the merged solution here https://github.com/SciML/DiffEqParamEstim.jl/blob/e99fb8233f48be99540dc9bf63ac226bb22c3c22/src/multiple_shooting_objective.jl#L58-L63

**Comment by Manas2030** [comment]
@yewalenikhil65  I wrote above comment reading your previous suggestion (before the edits) but for this one I think there would still be discontinuities. Here's a very naive example:

Suppose you are trying to predict the line, y = 100 at points x = 1 to 20 with grp_size/batchsize = 5
Suppose for first group your predictions lie above y = 100 with prediction at x = 1 being 100. So with some value of the parameters it's very much possible that for the next group predictions lie below y = 100 with prediction for x = 5 being 100.
This brings up a discontinuity.

Also there is nothing wrong with plotting discontinuities but I am not sure if it would be a better idea so honestly I was waiting for Chris's opinion on this 😅
Other than that it's a good suggestion, I don't see any or previous problems 😄 

@Vaibhavdixit02 I see. Do you think it would be a better idea if we add a parameter that let's the user decide what predictions they want? To be honest, it doesn't affect the training so anything should be fine and it's likely be used for only plotting purposes.

**Comment by ChrisRackauckas** [comment]
Multiple shooting is likely just for prediction, though it would be good to have a way to visualize it, otherwise it can be hard to see what's going on with the training. Indeed, DiffEqParamEstim.jl's multiple shooting was good, but you couldn't do more than define a loss, get parameters, and question what went wrong if it didn't fit.

Multiple shooting is easy to visualize:

https://d3i71xaburhd42.cloudfront.net/4284951012483c6914a7b1bfd7930863cb47f5cf/5-Figure2-1.png

For the discontinuities, just have two points exactly at `t` and it'll handle the discontinuities. 

https://diffeq.sciml.ai/stable/features/callback_functions/#Example-1:-Interventions-at-Preset-Times

That has tons of plots with discontinuities. So I think giving someone a good way to visualize the multiple shooting fit would be pretty good.

#### Larger comment

The larger vision here is that DiffEqParamEstim.jl was our first attempt at a parameter estimation library, but the autodiff enhancements led to DiffEqFlux.jl which is a lot more general. Multiple shooting is one of the last reasons to use DiffEqParamEstim.jl, because that library was based around defining loss functions while DiffEqFlux is all about the user defining their own loss functions. Except, multiple shooting is a fancy loss function. How do we make it easy to use a fancy loss function while keeping the flexibility of DiffEqFlux, things like having control over which variables are used when checking against the data, and switching between forward and reverse mode approaches? That's the question to answer here.

I think this PR has done a generally good job, though it needs to remove some of those `::Array` annotations. But, somehow adding a visualization to multiple shooting so it can be animated similar to the tutorials would be great for teaching and for diagnosing fitting issues.

**Comment by Manas2030** [comment]
> Needs to be `include`d at the top level, needs docs.

@ChrisRackauckas  What did you mean by top level? And for docs should I append it to https://diffeqflux.sciml.ai/dev/examples/local_minima/ or make a whole new tutorial? 😅

**Comment by ChrisRackauckas** [comment]
I mean, this file isn't part of the library yet. It hasn't been `include`d into the module. Separately, it needs docs and tests.

**Comment by Manas2030** [comment]
Oh okay, I see but where can I upload the code (as in which directory/location)? I am confused.

**Comment by Vaibhavdixit02** [comment]
@Manas2030 what Chris is saying is that you need to have the `multiple_shooting.jl` file `include`-ed in the src/DiffEqFlux.jl file where we define the module, please take a look at that file and that should clear things up for you.

>And for docs should I append it to https://diffeqflux.sciml.ai/dev/examples/local_minima/ or make a whole new tutorial?

I think, it can go in there, but I would like to see a convincing example of where you actually benefit with multiple shooting so let's get to that from a different PR with some discussion.

This PR needs a test. (Take a look at other tests it should be very easy to add one)

**Comment by Manas2030** [comment]
Ah, okay this makes it very clear. Thank you!

File: docs/src/examples/multiple_shooting.md [hunk 3]
```diff
@@ -0,0 +1,102 @@
+# Multiple Shooting
+
+In Multiple Shooting, the training data is split into overlapping intervals. The solver is then trained on individual intervals. If the end conditions of any interval co-incide with the initial conditions of the next immediate interval, then the joined/combined solution is same as solving on the whole dataset (without splitting).
```
Comment (ChrisRackauckas) [inline docs/src/examples/multiple_shooting.md:3]: Line lengths

File: docs/src/examples/multiple_shooting.md [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+function loss_function_param(ode_data, pred):: Float32
+	return sum(abs2, (ode_data .- pred))^2
+end
+
+function loss_neuralode(p)
+	return multiple_shoot(p, ode_data, tsteps, prob_param, loss_function_param, grp_size_param, loss_multiplier_param)
+end
+
+result_neuralode = DiffEqFlux.sciml_train(loss_neuralode, prob_neuralode.p,
+                                          ADAM(0.05), cb = callback,
```
Comment (ChrisRackauckas) [inline docs/src/examples/multiple_shooting.md:86]: you're missing the BFGS that comes after ADAM

File: docs/src/examples/multiple_shooting.md (hunk 3 above)
Comment (Manas2030) [reply docs/src/examples/multiple_shooting.md:3]: Do you mean that I should split this into different lines? (and also preferably write shorter lines in future)

Comment (ChrisRackauckas) [reply docs/src/examples/multiple_shooting.md:3]: Your text editor should show you the standard maximum line length. Break lines at that spot.
